runner/src/world/ini.rs

```diff
@@ -0,0 +1,341 @@
+}
+
+impl FileLike for File {
+    fn set_len(&mut self, size: u64) -> std::io::Result<()> {
+        File::set_len(self, size)
+    }
+}
+
+#[cfg(test)]
+impl FileLike for Cursor<&mut Vec<u8>> {
+    fn set_len(&mut self, size: u64) -> std::io::Result<()> {
+        self.get_mut().resize(size as usize, 0);
+        Ok(())
+    }
+}
+
+#[gml::bind]
+impl State {
+    #[gml::api]
+    pub fn ini_open(&mut self, filename: Symbol) {
+        // TODO: GML expects the filename to be a relative path for the same
+        // directory as the game, e.g. "game.ini".
+        self.current_path = Some(path_from_gml(filename).unwrap());
+    }
+
+    #[gml::api]
+    pub fn ini_close(&mut self) {
+        self.current_path = None;
+    }
+
+    // This is not the prettiest or fastest way to manipulate an INI file,
+    // but this is probably how the Windows API and Game Maker do it.
+    // Importantly, it is non-destructive.
+    fn read_or_write_key<F: FileLike>(
+        &mut self,
+        file: F,
+        section: &[u8],
+        key: &[u8],
+        new_value: Option<&[u8]>,
+    ) -> Option<Vec<u8>> {
+        let mut buffered_file = BufReader::new(file);
+
+        let mut found_section = false;
+        let mut buf: Vec<u8> = Vec::new();
+
+        let mut line_terminated = false;
+
+        while buffered_file.read_until(b'\n', &mut buf).is_ok() && buf.len() > 0 {
+            let line_end = buffered_file.stream_position().unwrap();
+            let line_start = line_end - buf.len() as u64;
+
+            // TODO: proper whitespace handling?
+            line_terminated = &buf[buf.len() - 2..] == b"\r\n";
+            if line_terminated {
+                buf.truncate(buf.len() - 2)
+            }
+
+            if buf.get(0) == Some(&b'[') && buf.get(buf.len() - 1) == Some(&b']') {
+                let old_found_section = found_section;
+                found_section = &buf[1..buf.len() - 1] == section;
+                if let Some(new_value) = new_value {
+                    if !found_section && old_found_section {
+                        // Didn't find the value in the section. Add it at the
+                        // end of the section.
+
+                        let mut file = buffered_file.into_inner();
+                        file.seek(SeekFrom::Start(line_start)).unwrap();
+                        buf.clear();
+                        file.read_to_end(&mut buf).unwrap();
+                        file.seek(SeekFrom::Start(line_start)).unwrap();
+                        file.write_all(key).unwrap();
+                        file.write_all(b"=").unwrap();
+                        file.write_all(new_value).unwrap();
+                        file.write_all(b"\r\n").unwrap();
+                        file.write_all(&buf).unwrap();
+                        return None;
+                    }
+                }
+            }
+
+            if found_section && &buf[0..key.len()] == key && buf.get(key.len()) == Some(&b'=') {
```

Comment:
It looks like this panics when the current line is shorter than the key we're looking for? E.g. searching for `loooooong` in this file:

```rust
[section]
short=0
loooooong=1
```

Reply:
Oh, somehow I thought ranges on slices were kind about this… maybe it's just ones where it's partially unbounded.

That explains why `Vec::get` exists for ranges I guess.

Reply:
I fixed this now, and also made the section and key matching a bit prettier.